File: src/collections_and_containers/cpp/vector/test/test_vector.cpp
#include "vector.h"
#include "gtest/gtest.h"

TEST(VectorTest, EmptyVector) {
  Vector<int> vector;
  ASSERT_TRUE(vector.empty());
}

TEST(VectorTest, Clear) {
  Vector<int> vector;
  vector.push_back(1);
  vector.push_back(2);
  vector.push_back(3);
  EXPECT_EQ(vector.size(), 3);

  vector.clear();
  ASSERT_TRUE(vector.empty());
}

TEST(VectorTest, AddingMultipleElements) {
  Vector<int> vector;
  vector.push_back(1);
  vector.push_back(2);
  vector.push_back(3);

  EXPECT_EQ(vector.size(), 3);
  EXPECT_EQ(vector.front(), 1);
  EXPECT_EQ(vector.back(), 3);
}

TEST(VectorTest, Front) {
  Vector<int> vector;
  vector.push_back(1);
  vector.push_back(2);
  vector.push_back(3);

  EXPECT_EQ(vector.front(), 1);
}

TEST(VectorTest, Back) {
  Vector<int> vector;
  vector.push_back(1);
  vector.push_back(2);
  vector.push_back(3);

  EXPECT_EQ(vector.back(), 3);
}

TEST(VectorTest, Resize) {
  Vector<int> vector;
  vector.resize(1111);
  EXPECT_EQ(vector.size(), 1111);
}

TEST(VectorTest, Indexing) {
  Vector<int> vector;
  vector.push_back(1);
  vector.push_back(2);
  vector.push_back(3);

  EXPECT_EQ(vector[0], 1);
  EXPECT_EQ(vector[1], 2);
  EXPECT_EQ(vector[2], 3);
}

TEST(VectorTest, CopyConstructor) {
  Vector<int> vector;
  vector.push_back(1);
  vector.push_back(2);
  vector.push_back(3);

  Vector<int> vector2(vector);
  EXPECT_EQ(vector2.size(), 3);
  EXPECT_EQ(vector2[0], 1);
  EXPECT_EQ(vector2[1], 2);
  EXPECT_EQ(vector2[2], 3);
}

TEST(VectorTest, AssignmentOperator) {
  Vector<int> vector;
  vector.push_back(1);
  vector.push_back(2);
  vector.push_back(3);

  Vector<int> vector2;
  vector2 = vector;
  EXPECT_EQ(vector2.size(), 3);
  EXPECT_EQ(vector2[0], 1);
  EXPECT_EQ(vector2[1], 2);
  EXPECT_EQ(vector2[2], 3);
}

TEST(VectorTest, Iterator) {
  Vector<int> vector;
  vector.push_back(1);
  vector.push_back(2);
  vector.push_back(3);

  Vector<int>::iterator it = vector.begin();
  EXPECT_EQ(*it, 1);
  ++it;
  EXPECT_EQ(*it, 2);
  ++it;
  EXPECT_EQ(*it, 3);
  ++it;
  EXPECT_EQ(it, vector.end());
}

TEST(VectorTest, AcessingIllegalIndex) {
  Vector<int> vector;
  vector.push_back(1);
  vector.push_back(2);
  vector.push_back(3);

  ASSERT_THROW(vector[3], std::out_of_range);
  ASSERT_THROW(vector[-1], std::out_of_range);
}

File: src/collections_and_containers/python/vector/test/test_vector.py
from src.vector import Vector
from unittest import TestCase
import unittest


class TestVector(TestCase):
    def test_empty_vector(self):
        vector = Vector()
        self.assertTrue(vector.empty())

    def test_adding_multiple_elements(self):
        vector = Vector()
        vector.push_back(1)
        vector.push_back(2)
        vector.push_back(3)
        self.assertEqual(vector.size(), 3)

        self.assertEqual(vector.pop_back(), 3)
        self.assertEqual(vector.pop_back(), 2)
        self.assertEqual(vector.pop_back(), 1)

        self.assertTrue(vector.empty())

        # adding the same elements in different order
        vector.push_back(3)
        vector.push_back(1)
        vector.push_back(2)

        self.assertEqual(vector.pop_back(), 2)
        self.assertEqual(vector.pop_back(), 1)
        self.assertEqual(vector.pop_back(), 3)

    def test_front(self):
        vector = Vector()
        vector.push_back(1)
        vector.push_back(2)
        vector.push_back(3)

        self.assertEqual(vector.front(), 1)

    def test_front_when_empty(self):
        vector = Vector()
        self.assertRaises(IndexError, vector.front)

    def test_back(self):
        vector = Vector()
        vector.push_back(1)
        vector.push_back(2)
        vector.push_back(3)

        self.assertEqual(vector.back(), 3)

    def test_back_when_empty(self):
        vector = Vector()
        self.assertRaises(IndexError, vector.back)

    def test_pop_back_when_empty(self):
        vector = Vector()
        self.assertRaises(IndexError, vector.pop_back)

    def test_resize(self):
        vector = Vector()
        vector.push_back(1)
        vector.push_back(2)
        vector.push_back(3)
        vector.resize(2)
        self.assertEqual(vector.size(), 2)
        self.assertEqual(vector[0], 1)
        self.assertEqual(vector[1], 2)
        self.assertRaises(IndexError, vector.__getitem__, 2)

    def test_resize_when_empty(self):
        vector = Vector()
        vector.resize(2)
        self.assertEqual(vector.size(), 2)
        self.assertEqual(vector[0], None)
        self.assertEqual(vector[1], None)
        self.assertRaises(IndexError, vector.__getitem__, 2)

    def test_clear(self):
        vector = Vector()
        vector.push_back(1)
        vector.push_back(2)
        vector.push_back(3)
        vector.clear()
        self.assertTrue(vector.empty())
